compactor/src/handler.rs

```diff
@@ -414,6 +406,167 @@ async fn compact_hot_partitions(compactor: Arc<Compactor>) -> usize {
+            continue;
+        }
+        let parquet_files_for_compaction = parquet_files_for_compaction.unwrap();
+
+        // Return only files under the remaining_budget_bytes that should be compacted
+        let to_compact = parquet_file_filtering::filter_hot_parquet_files(
+            partition.clone(),
+            parquet_files_for_compaction,
+            remaining_budget_bytes,
+            columns,
+            &compactor.parquet_file_candidate_gauge,
+            &compactor.parquet_file_candidate_bytes,
+        );
+
+        // --------------------------------------------------------------------
+        // 3. Check the compactable status and act provide the right action
+        match to_compact.filter_result() {
+            FilterResult::NothingToCompact => {
+                continue;
+            }
+            FilterResult::ErrorEstimatingBudget => {
+                warn!(
+                    ?partition_id,
+                    ?table_id,
+                    "hot compaction is skipped due to error in estimating compacting memory"
+                );
+                // todo: add this partition and its info into a new catalog table
+                // https://github.com/influxdata/influxdb_iox/issues/5458
+                continue;
+            }
+            FilterResult::OverBudget => {
+                if remaining_budget_bytes < compactor.config.memory_budget_bytes {
+                    // Add this partition back the end of the list to compact it with full budget later
+                    candidates.push(partition);
+                } else {
+                    // Even with max budget, we cannot compact a bit of this partition, log it
+                    warn!(
+                        ?partition_id,
+                        ?table_id,
+                        "hot compaction is skipped due to over memory budget"
+                    );
+                    // todo: add this partition and its info into a new catalog table
+                    // https://github.com/influxdata/influxdb_iox/issues/5458
+                }
+                continue;
+            }
+            FilterResult::Proceeed => {
+                remaining_budget_bytes -= to_compact.budget_bytes();
+                parallel_compacting_candidates.push(to_compact);
+            }
+        }
+
+        // --------------------------------------------------------------------
+        // 4. Almost hitting max budget or no more candidates, let compact parallel_compacting_candidates
+        if (remaining_budget_bytes < (compactor.config.memory_budget_bytes as f64 * 0.5) as u64)
+            || (candidates.is_empty() && !parallel_compacting_candidates.is_empty())
+        {
+            let partitions = parallel_compacting_candidates
+                .drain(0..)
+                .collect::<Vec<_>>();
+            compact_hot_partitions_in_parallel(Arc::clone(&compactor), partitions).await;
```

Comment:
I was confused about what this `drain` `collect` was doing because it looked like a no-op, then I tried removing it and saw that this is emptying `parallel_compacting_candidates` so that it will be empty for the next iteration of the candidate loop. I think this would be clearer and possibly faster as:

```suggestion
            compact_hot_partitions_in_parallel(Arc::clone(&compactor), parallel_compacting_candidates).await;
            parallel_compacting_candidates = Vec::with_capacity(candidates.len());
```

The ownership of the `Vec` in `parallel_compacting_candidates` gets transferred to `compact_hot_partitions_in_parallel`, but the binding gets reinitialized before it's used again.